src/algebra/category/RingModPair/basic.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebra.category.Ring.basic
+import algebra.category.Module.change_of_rings
+
+/-!
+# The category of ring-module pair
+
+`RingModPair` is the category of pairs of the shape `(R, M)` where `R` is
+a ring and `M` is an `R`-module. A morphism between `M1 = (R, M)`
+and `M2 = (S, N)` is a pair of morphism `(f, g)` where `f : R ⟶ S` is a ring
+homomorphism and `g : M ⟶ f* N` is a module homomorphism (linear map).
+-/
+
+namespace category_theory
+
+open category_theory.Module category_theory.limits
+
+section RingModPair
+
+universes u v
+
+/--
+A ring-module pair is a pair `(R, M)` such that `R : Ring` and `M` is an `R`-module.
+-/
+@[nolint check_univs]
+structure RingModPair : Type (max (v+1) (u+1)) :=
+(ring : Ring.{u})
+(mod : Module.{v u} ring)
+
+instance : inhabited RingModPair :=
+{ default := ⟨⟨punit⟩, ⟨punit⟩⟩ }
+
+namespace RingModPair
+
+/--
+A morphism between `M1 = (R, M)`
+and `M2 = (S, N)` is a pair of morphism `(f, g)` where `f : R ⟶ S` is a ring
+homomorphism and `g : M ⟶ f* N` is a module homomorphism (linear map)
+-/
+def hom (P Q : RingModPair) : Type (max v u) :=
+Σ (ring_hom : P.ring ⟶ Q.ring),
+  P.mod ⟶ (category_theory.Module.restrict_scalars ring_hom).obj Q.mod
```

Comment:
I wonder if you should use semilinear maps instead. This at least avoids the need to define `id` and `comp` by hand below.